### PR TITLE
Adding exercises to workout collections

### PR DESCRIPTION
**Pull request by craig-a-lee**
User can now create custom workouts
Used singleton pattern to reduce number of API calls when app is in use. 
App now fetches all exercises one time when the app is opened rather than each time the search tab is selected
Queries into the dictionary obtained from the api are now done locally instead of via api access

### CODE REVIEW

File: Wellthier/Cells/WorkoutTableViewCell.m [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+- (void)awakeFromNib {
+    [super awakeFromNib];
+    // Initialization code
+}
+
+- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
+    [super setSelected:selected animated:animated];
+
+    // Configure the view for the selected state
+}
```
Comment (vijaykanse) [inline Wellthier/Cells/WorkoutTableViewCell.m:18]: Please delete the overrides if not needed.

File: Wellthier/ExerciseCommonManager.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+//
+//  Created by Craig Lee on 7/18/22.
+//
+
+#import <Foundation/Foundation.h>
+#import "Exercise.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface ExerciseCommonManager : NSObject //change to exercise shared manager
```
Comment (vijaykanse) [inline Wellthier/ExerciseCommonManager.h]: May be correct this one.

File: Wellthier/Models/SearchCell.h [hunk 3]
```diff
@@ -10,8 +10,8 @@
 NS_ASSUME_NONNULL_BEGIN
 
 @interface SearchCell : UICollectionViewCell
-@property (weak, nonatomic) IBOutlet UIButton *button;
 
+@property (weak, nonatomic) IBOutlet UIButton *button;
```
Comment (vijaykanse) [inline Wellthier/Models/SearchCell.h:14]: nit: may be rename it addWorkoutButton

File: Wellthier/Models/Workout.h [hunk 4]
```diff
@@ -20,6 +21,8 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, strong) NSMutableArray *exercises;
 
 + (void) postUserWorkout: ( UIImage * _Nullable )image withTitle: ( NSString * _Nullable )title withCompletion: (PFBooleanResultBlock  _Nullable)completion;
++ (void) updateUserWorkout: ( Workout * _Nullable )workout withExercise: ( Exercise * _Nullable )exercise withCompletion: (PFBooleanResultBlock  _Nullable)completion;
++ (void) deleteExerciseFromWorkout: ( Workout * _Nullable )workout withExercise: ( Exercise * _Nullable )exercise withCompletion: (PFBooleanResultBlock  _Nullable)completion;
```
Comment (vijaykanse) [inline Wellthier/Models/Workout.h]: avoid additional space in arguments
+ (void) deleteExerciseFromWorkout: (Workout * _Nullable)workout withExercise: (Exercise * _Nullable)exercise withCompletion: (PFBooleanResultBlock  _Nullable)completion;

File: Wellthier/Models/Workout.m [hunk 5]
```diff
@@ -32,6 +33,16 @@ + (void) postUserWorkout: ( UIImage * _Nullable )image withTitle: ( NSString * _
         [newWorkout saveInBackgroundWithBlock: completion];
 }
 
++ (void) updateUserWorkout: ( Workout * _Nullable )workout withExercise: ( Exercise * _Nullable )exercise withCompletion: (PFBooleanResultBlock  _Nullable)completion {
```
Comment (vijaykanse) [inline Wellthier/Models/Workout.m]: same here

File: Wellthier/View Controllers/AddToWorkoutViewController.h [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+//
+//  Created by Craig Lee on 7/18/22.
+//
+
+#import "ViewController.h"
+#import "Workout.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface AddToWorkoutViewController : ViewController
```
Comment (vijaykanse) [inline Wellthier/View Controllers/AddToWorkoutViewController.h:13]: What is use of ViewController class ?

File: Wellthier/View Controllers/AddToWorkoutViewController.m [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+@property (weak, nonatomic) IBOutlet UITableView *tableView;
+
+@end
+
+@implementation AddToWorkoutViewController
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    // Do any additional setup after loading the view.
```
Comment (vijaykanse) [inline Wellthier/View Controllers/AddToWorkoutViewController.m]: please remove

File: Wellthier/View Controllers/AddToWorkoutViewController.m [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+- (void)getWorkouts {
+    PFQuery *workoutQuery = [Workout query];
+    [workoutQuery includeKey:@"author"];
+    [workoutQuery whereKey:@"author" equalTo:PFUser.currentUser];
+    [workoutQuery whereKey:@"title" notEqualTo:@"Liked Exercises"];
+    [workoutQuery findObjectsInBackgroundWithBlock:^(NSArray<Workout *> * _Nullable workouts, NSError * _Nullable error) {
+        if (workouts) {
+            // do something with the data fetched
```
Comment (vijaykanse) [inline Wellthier/View Controllers/AddToWorkoutViewController.m]: Please remove

File: Wellthier/View Controllers/AddToWorkoutViewController.m [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    WorkoutTableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"WorkoutTableViewCell"];
+    Workout *workout = self.filteredWorkouts[indexPath.row];
+    cell.workoutImageView.file = workout[@"image"];
+    [cell.workoutImageView loadInBackground];
+    PFUser *user = workout[@"author"];
+    cell.workoutName.text = workout.title;
+    cell.workoutAuthorInfo.text = user.username;
+    if ([cell isSelected]) {
+        cell.accessoryType = UITableViewCellAccessoryCheckmark;
+    }
```
Comment (vijaykanse) [inline Wellthier/View Controllers/AddToWorkoutViewController.m]:
```suggestion
if ([cell isSelected]) {
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
    }
    else {
      cell. accessoryType = UITableViewCellAccessoryNone;
    }
```

File: Wellthier/View Controllers/NewWorkoutViewController.h [hunk 10]
```diff
@@ -9,8 +9,15 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+@protocol NewWorkoutViewControllerDelegate
```
Comment (vijaykanse) [inline Wellthier/View Controllers/NewWorkoutViewController.h:12]: Very Nice

File: Wellthier/View Controllers/NewWorkoutViewController.m [hunk 11]
```diff
@@ -56,9 +55,40 @@ - (IBAction)didTapUploadPic:(id)sender {
...
+    
+    // Get the image captured by the UIImagePickerController
+    UIImage *originalImage = info[UIImagePickerControllerOriginalImage];
 
-- (IBAction)didTapCreate:(id)sender {
+    // Do something with the images (based on your use case)
+    CGSize size = CGSizeMake(167, 167);
+    self.workoutImage.image = [self resizeImage:originalImage withSize:size];
+    
+    // Dismiss UIImagePickerController to go back to your original view controller
```
Comment (vijaykanse) [inline Wellthier/View Controllers/NewWorkoutViewController.m]: Please remove

File: Wellthier/View Controllers/WorkoutViewController.h [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+@interface WorkoutViewController : ViewController
+
+@property (weak, nonatomic) IBOutlet UILabel *workoutTitle;
+@property (weak, nonatomic) IBOutlet UILabel *author;
+@property (strong, nonatomic) NSArray *arrayOfExercises;
+@property (strong, nonatomic) NSArray *filteredExercises;
+@property (strong, nonatomic) Workout *detailWorkout;
+@property (weak, nonatomic) IBOutlet PFImageView *workoutImageView;
```
Comment (vijaykanse) [inline Wellthier/View Controllers/WorkoutViewController.h:22]: I would like to keep these properties in consistent order for ex:
(nonatomic, weak)
(nonatomic, strong)

File: Wellthier/View Controllers/WorkoutViewController.m [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+@implementation WorkoutViewController
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    self.tableView.delegate = self;
+    self.tableView.dataSource = self;
+    self.searchBar.delegate = self;
+    self.workoutTitle.text = self.detailWorkout.title;
+    // Do any additional setup after loading the view.
```
Comment (vijaykanse) [inline Wellthier/View Controllers/WorkoutViewController.m]: Please remove

File: Wellthier/View Controllers/WorkoutViewController.m [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    self.workoutImageView.file = self.detailWorkout[@"image"];
+    PFUser *user = self.detailWorkout[@"author"];
+    self.author.text = user[@"displayName"];
+    [self getExercises];
+}
+
+- (void) getExercises {
+    self.arrayOfExercises = [NSMutableArray new];
+    self.filteredExercises = [NSMutableArray new];
```
Comment (vijaykanse) [inline Wellthier/View Controllers/WorkoutViewController.m]: We may not need these two lines

File: Wellthier/View Controllers/WorkoutViewController.m [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    cell.name.animationCurve = UIViewAnimationCurveEaseIn;
+    cell.name.fadeLength = 10.0;
+    cell.name.scrollDuration = 3.0;
+    cell.bodyPart.text = [ex.bodyPart capitalizedString];
+    return cell;
+}
+
+#pragma mark - Navigation
+
+// In a storyboard-based application, you will often want to do a little preparation before navigation
```
Comment (vijaykanse) [inline Wellthier/View Controllers/WorkoutViewController.m]: please remove

File: Wellthier/View Controllers/WorkoutViewController.m [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+    // Get the new view controller using [segue destinationViewController].
+    // Pass the selected object to the new view controller.
```
Comment (vijaykanse) [inline Wellthier/View Controllers/WorkoutViewController.m]: please remove

File: Wellthier/View Controllers/AddToWorkoutViewController.h (hunk 6 above)
Comment (craig-a-lee) [reply Wellthier/View Controllers/AddToWorkoutViewController.h:13]: unclear what this comment means

File: Wellthier/View Controllers/WorkoutViewController.h (hunk 12 above)
Comment (craig-a-lee) [reply Wellthier/View Controllers/WorkoutViewController.h:22]: so is the current order fine?

**Pull request merged**